test/expr/unary/keypath/keypath.swift

```diff
@@ -230,9 +230,9 @@ func testDisembodiedStringInterpolation(x: Int) {
 
 func testNoComponents() {
   let _: KeyPath<A, A> = \A // expected-error{{must have at least one component}}
-  let _: KeyPath<C, A> = \C // expected-error{{must have at least one component}} expected-error{{}}
+  let _: KeyPath<C, A> = \C // expected-error{{must have at least one component}}
   // expected-error@-1 {{generic parameter 'T' could not be inferred}}
-  // expected-error@-2 {{cannot convert value of type 'KeyPath<Root, Value>' to specified type 'KeyPath<C<T>, A>'}}
+  let _: KeyPath<A, C> = \A // expected-error{{must have at least one component}}
```

Comment:
Hm, only concern I have regarding this case is it seems like we’d form a _valid_ solution and apply it although the source code is incorrect... it seems like at least having a contextual mismatch is such case is valid.

Reply:
Do you mean the actual diagnostic message in this case `KeyPath<A, A> to KeyPath<A, C<T>`? 
Because the fixes are all still recorded(here we are just assuming as diagnosed before emitting message)... but in this specific case the solution being found as "best solution" is the one which diagnoses a generic mismatch

Reply:
@xedin I was taking a deeper look at this, this case produces 3 solutions(ambiguos handled by `diagnoseAmbiguityWithContextualType`)... but as far as I could see none of them were being applied :)

Reply:
Sorry I just finally got to look at this again - if we have 3 solutions where why there is no ambiguity diagnostic?

Reply:
Is not an ambiguity diagnostic but as far as I could see this case is handled by `diagnoseAmbiguityWithFixes` particularly `diagnoseAmbiguityWithContextualType` where all fixes produce the same result across all solutions, so they are diagnosed as a single failure, that being diagnosing one of the fixes. 
Here https://github.com/apple/swift/blob/142aa5f9996954c32a311932e6cbd7652414f008/lib/Sema/ConstraintSystem.cpp#L3119

Reply:
That makes sense but my actual question is - why there is only "must have at least one component" diagnostic here and not "cannot convert to specified type" as well?

Reply:
Ah, do you think it makes sense a contextual failure here too in this last case?

Reply:
Yeah, I think it makes sense to report that or report “at least one component” error through a fix to make sure that error is always recorded and reported.

Reply:
Right, I have to think a bit about this then, because then we wouldn't rely on this being already diagnosed on `TypeCheckConstraints`, but always record a fix new for `must have at least one component` on invalid cases? Did I understand it correctly? If it is that the idea, we would record a fix that wouldn't be diagnosed, or do we move the diagnostic from `TypeCheckConstraints` to always diagnose in the fix?

Reply:
The idea is to avoid diagnosing it early and instead diagnose it via a fix and pretend that there a viable component represented by a fresh type variable.

Reply:
Ah Ok, I'll get back on that ASAP. Thank you :)

Reply:
No worries!

Reply:
@xedin All adjusted, let me know what you think :)